pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,452 @@
+	if jobNeedsSync {
+		active = jm.manageJob(activePods, successful, unsuccessful, &job)
+	}
+	completions := successful
+	if job.Spec.Template.Spec.RestartPolicy == api.RestartPolicyNever {
+		completions += unsuccessful
+	}
+	if completions == *job.Spec.Completions {
+		job.Status.Conditions = append(job.Status.Conditions, newCondition())
+	}
+
+	// no need to update the job if the status hasn't changed since last time
+	if job.Status.Active != active || job.Status.Successful != successful || job.Status.Unsuccessful != unsuccessful {
+		job.Status.Active = active
+		job.Status.Successful = successful
+		job.Status.Unsuccessful = unsuccessful
+
+		if err := jm.updateHandler(&job); err != nil {
+			glog.V(2).Infof("Failed to update job %v, requeuing", job.Name)
+			jm.enqueueController(&job)
+		}
+	}
+	return nil
+}
+
+func newCondition() experimental.JobCondition {
+	return experimental.JobCondition{
+		Type:               experimental.JobComplete,
+		Status:             api.ConditionTrue,
+		LastProbeTime:      util.Now(),
+		LastTransitionTime: util.Now(),
+	}
+}
+
+func getStatus(jobKey string, restartPolicy api.RestartPolicy, pods []api.Pod) (successful, unsuccessful int) {
+	successful = filterPods(pods, api.PodSucceeded)
+	if restartPolicy == api.RestartPolicyNever {
+		unsuccessful = filterPods(pods, api.PodFailed)
+	}
+	return
+}
+
+func (jm *JobManager) manageJob(activePods []*api.Pod, successful, unsuccessful int, job *experimental.Job) int {
+	active := len(activePods)
+	parallelism := *job.Spec.Parallelism
+	jobKey, err := controller.KeyFunc(job)
+	if err != nil {
+		glog.Errorf("Couldn't get key for job %#v: %v", job, err)
+		return 0
+	}
+
+	if active > parallelism {
+		diff := active - parallelism
+		jm.expectations.ExpectDeletions(jobKey, diff)
+		glog.V(2).Infof("Too many pods running job %q, need %d, deleting %d", jobKey, parallelism, diff)
+		// Sort the pods in the order such that not-ready < ready, unscheduled
+		// < scheduled, and pending < running. This ensures that we delete pods
+		// in the earlier stages whenever possible.
+		sort.Sort(controller.ActivePods(activePods))
+
+		active -= diff
+		wait := sync.WaitGroup{}
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func(ix int) {
+				defer wait.Done()
+				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name); err != nil {
+					// Decrement the expected number of deletes because the informer won't observe this deletion
+					glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q", jobKey)
+					jm.expectations.DeletionObserved(jobKey)
+					util.HandleError(err)
+					active++
+				}
+			}(i)
+		}
+		wait.Wait()
+
+	} else if active < parallelism {
+		// how many executions are left to run
+		diff := *job.Spec.Completions - successful
+		// for RestartPolicyNever we need to count unsuccessful pods as well
+		if job.Spec.Template.Spec.RestartPolicy == api.RestartPolicyNever {
+			diff -= unsuccessful
```

Comment:
Or make this match the doc and the design, and we can revisit adding failed completion after further design discussion.